compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -476,6 +476,56 @@ impl<'tcx> ThirBuildCx<'tcx> {
                 ExprKind::RawBorrow { mutability, arg: self.mirror_expr(arg) }
             }
 
+            // Make `&pin mut $expr` and `&pin const $expr` into
+            // `Pin { __pointer: &mut { $expr } }` and `Pin { __pointer: &$expr }`.
+            hir::ExprKind::AddrOf(hir::BorrowKind::Pin, mutbl, arg) => match expr_ty.kind() {
+                &ty::Adt(adt_def, args)
+                    if tcx.is_lang_item(adt_def.did(), rustc_hir::LangItem::Pin) =>
+                {
+                    let ty = args.type_at(0);
+                    let arg_ty = self.typeck_results.expr_ty(arg);
+                    let mut arg = self.mirror_expr(arg);
+                    // for `&pin mut $place`, move the place `$place` to ensure it will not be used afterwards
```

Comment:
```suggestion
                    // For `&pin mut $place`, move the place `$place` to ensure
                    // it will not be used afterwards.
```